js/ui/selection/selection.strategy.deferred.js

```diff
@@ -179,37 +179,39 @@ export default SelectionStrategy.inherit({
         });
     },
 
-    _addSelectionFilter: function(isDeselect, filter, isSelectAll) {
+    _addSelectionFilter: function(isDeselect, filter, isSelectAll, skipFilter) {
         const that = this;
         const currentFilter = isDeselect ? ['!', filter] : filter;
         const currentOperation = isDeselect ? 'and' : 'or';
         let needAddFilter = true;
         let selectionFilter = that.options.selectionFilter || [];
 
         selectionFilter = that._denormalizeFilter(selectionFilter);
+        if(selectionFilter && selectionFilter.length && !skipFilter) {
 
-        if(selectionFilter && selectionFilter.length) {
             const removedIndex = that._removeSameFilter(selectionFilter, filter, isDeselect, isSelectAll);
             const filterIndex = that._removeSameFilter(selectionFilter, filter, !isDeselect);
 
             const shouldCleanFilter =
-                isDeselect &&
-                (removedIndex !== -1 || filterIndex !== -1) &&
-                this._isOnlyNegativeFiltersLeft(selectionFilter);
+                    isDeselect &&
+                    (removedIndex !== -1 || filterIndex !== -1) &&
+                    this._isOnlyNegativeFiltersLeft(selectionFilter);
 
             if(shouldCleanFilter) {
                 selectionFilter = [];
             }
 
+
             const isKeyOperatorsAfterRemoved = this._isKeyFilter(filter) && this._hasKeyFiltersOnlyStartingFromIndex(selectionFilter, filterIndex);
 
             needAddFilter = filter.length && !isKeyOperatorsAfterRemoved;
+        }
 
-            if(needAddFilter) {
-                selectionFilter = that._addFilterOperator(selectionFilter, currentOperation);
-            }
+        if(needAddFilter) {
+            selectionFilter = that._addFilterOperator(selectionFilter, currentOperation);
         }
 
+
         if(needAddFilter) {
             selectionFilter.push(currentFilter);
         }
```

Comment:
Let's combine these conditions:

```suggestion
        if(needAddFilter) {
            selectionFilter = that._addFilterOperator(selectionFilter, currentOperation);
            selectionFilter.push(currentFilter);
        }
```